### PR TITLE
returns reshares in API

### PR DESCRIPTION
**Pull request by ArtificialOwl**
(#14112) Returns reshares to the API

### CODE REVIEW

**Review by tobiasKaminsky — APPROVED** [review]
Works for me :+1:

**Comment by MorrisJobke** [comment]
@daita This breaks the sharing tests.

**Comment by tobiasKaminsky** [comment]
@daita for me it seems as if "path" is wrong:
- have a file /foo with 
	- shared with you by user0
	- shared to user1
	- shared to user3
- be in "/"
- query /nc/ocs/v2.php/apps/files_sharing/api/v1/shares?path=%2F&reshares=true&subfiles=true
- get 
```
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>200</statuscode>
  <message>OK</message>
 </meta>
 <data>
  <element>
   <id>401</id>
   <share_type>0</share_type>
   <uid_owner>user0</uid_owner>
   <displayname_owner>user0</displayname_owner>
   <permissions>15</permissions>
   <stime>1549621868</stime>
   <parent/>
   <expiration/>
   <token/>
   <uid_file_owner>user0</uid_file_owner>
   <note></note>
   <label/>
   <displayname_file_owner>user0</displayname_file_owner>
   <path>/</path>
   <item_type>folder</item_type>
   <mimetype>httpd/unix-directory</mimetype>
   <storage_id>home::user2</storage_id>
   <storage>28</storage>
   <item_source>75844</item_source>
   <file_source>75844</file_source>
   <file_parent>75843</file_parent>
   <file_target>/foo</file_target>
   <share_with>user1</share_with>
   <share_with_displayname>user1</share_with_displayname>
   <mail_send>0</mail_send>
   <hide_download>0</hide_download>
  </element>
  <element>
   <id>402</id>
   <share_type>0</share_type>
   <uid_owner>user0</uid_owner>
   <displayname_owner>user0</displayname_owner>
   <permissions>31</permissions>
   <stime>1549621887</stime>
   <parent/>
   <expiration/>
   <token/>
   <uid_file_owner>user0</uid_file_owner>
   <note></note>
   <label/>
   <displayname_file_owner>user0</displayname_file_owner>
   <path>/</path>
   <item_type>folder</item_type>
   <mimetype>httpd/unix-directory</mimetype>
   <storage_id>home::user2</storage_id>
   <storage>28</storage>
   <item_source>75844</item_source>
   <file_source>75844</file_source>
   <file_parent>75843</file_parent>
   <file_target>/foo</file_target>
   <share_with>user2</share_with>
   <share_with_displayname>user2</share_with_displayname>
   <mail_send>0</mail_send>
   <hide_download>0</hide_download>
  </element>
  <element>
   <id>403</id>
   <share_type>0</share_type>
   <uid_owner>user0</uid_owner>
   <displayname_owner>user0</displayname_owner>
   <permissions>31</permissions>
   <stime>1549622481</stime>
   <parent/>
   <expiration/>
   <token/>
   <uid_file_owner>user0</uid_file_owner>
   <note></note>
   <label/>
   <displayname_file_owner>user0</displayname_file_owner>
   <path>/</path>
   <item_type>folder</item_type>
   <mimetype>httpd/unix-directory</mimetype>
   <storage_id>home::user2</storage_id>
   <storage>28</storage>
   <item_source>75844</item_source>
   <file_source>75844</file_source>
   <file_parent>75843</file_parent>
   <file_target>/foo</file_target>
   <share_with>user3</share_with>
   <share_with_displayname>user3</share_with_displayname>
   <mail_send>0</mail_send>
   <hide_download>0</hide_download>
  </element>
 </data>
</ocs>
```

With master (on a file I shared with user admin): 
```
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>200</statuscode>
  <message>OK</message>
 </meta>
 <data>
  <element>
   <id>423</id>
   <share_type>0</share_type>
   <uid_owner>user2</uid_owner>
   <displayname_owner>user2</displayname_owner>
   <permissions>31</permissions>
   <stime>1553085500</stime>
   <parent/>
   <expiration/>
   <token/>
   <uid_file_owner>user2</uid_file_owner>
   <note></note>
   <label/>
   <displayname_file_owner>user2</displayname_file_owner>
   <path>/share</path>
   <item_type>folder</item_type>
   <mimetype>httpd/unix-directory</mimetype>
   <storage_id>home::user2</storage_id>
   <storage>28</storage>
   <item_source>77824</item_source>
   <file_source>77824</file_source>
   <file_parent>75844</file_parent>
   <file_target>/share</file_target>
   <share_with>admin</share_with>
   <share_with_displayname>admin</share_with_displayname>
   <mail_send>0</mail_send>
   <hide_download>0</hide_download>
  </element>
 </data>
</ocs>
```

**Comment by tobiasKaminsky** [comment]
Another question: should this really return the same user? 
I can filter this on client side, but is not useful at all, or?

**Comment by daita** [comment]
@MorrisJobke @ChristophWurst  any reason that most of the errors are from the TwoFactorBackupCodes app ?

**Comment by MorrisJobke** [comment]
> @MorrisJobke @ChristophWurst any reason that most of the errors are from the TwoFactorBackupCodes app ?

Ignore the risky tests, those are just warnings that there was no check (due to the environment) - the problem here are the failures.

**Comment by daita** [comment]
@MorrisJobke  How can I just run one scenario from the tests ?

**Comment by MorrisJobke** [comment]
> @MorrisJobke How can I just run one scenario from the tests ?

That should do the trick:

```
cd tests acceptance
./run.sh features/app-files-sharing-link.feature
```

**Comment by daita** [comment]
@MorrisJobke could not point the source of the issue but my guess is that a rebase (done) should do the trick

**Comment by daita** [comment]
@faily-bot still returns error, but nothing on my own tests, rebasing.

```
maxence@stealth:~/PhpstormProjects/server$ ./autotest.sh sqlite apps/files_sharing/tests/ApiTest.php 
Using PHP executable /usr/bin/php
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for sqlite testing on local storage ...
Installing ....
Nextcloud was successfully installed
Testing with sqlite ...
/usr/bin/phpunit --configuration phpunit-autotest.xml --coverage-clover autotest-clover-sqlite.xml --coverage-html coverage-html-sqlite --log-junit autotest-results-sqlite.xml ../apps/files_sharing/tests/ApiTest.php 
PHPUnit 6.5.5 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.15-0ubuntu0.18.04.2 with Xdebug 2.6.0
Configuration: /home/maxence/PhpstormProjects/server/tests/phpunit-autotest.xml

...................................                               35 / 35 (100%)

Time: 29.39 seconds, Memory: 56.00MB

OK (35 tests, 76 assertions)
```

**Comment by daita** [comment]
@MorrisJobke : looks green to me, some hiccup

**Comment by tobiasKaminsky** [comment]
@daita have you fixed 
https://github.com/nextcloud/server/pull/14605#issuecomment-474811355 and https://github.com/nextcloud/server/pull/14605#issuecomment-474818790?

**Comment by daita** [comment]
The first one is fixed and was the reason of the tests issue.
I don't understand the second one

File: lib/private/Share20/DefaultShareProvider.php [hunk 1]
```diff
@@ -609,7 +609,7 @@ public function getSharesInFolder($userId, Folder $node, $reshares) {
 	 */
 	public function getSharesBy($userId, $shareType, $node, $reshares, $limit, $offset) {
 		$qb = $this->dbConn->getQueryBuilder();
-		$qb->select('*')
+		$qb->selectDistinct('*')
```
Comment (MorrisJobke) [inline lib/private/Share20/DefaultShareProvider.php]: cc @rullzer

**Comment by tobiasKaminsky** [comment]
> The first one is fixed and was the reason of the tests issue.

:+1: working

> I don't understand the second one

- share a file /A from user admin to user tobi
- query http://localhost/nc/ocs/v1.php/apps/files_sharing/api/v1/shares?format=json&reshares=true&subfiles=true&path=%2F
- see
```
 {
        "id": "364",
        "share_type": 0,
        "uid_owner": "admin",
        "displayname_owner": "admin",
        "permissions": 31,
        "stime": 1540475407,
        "parent": null,
        "expiration": null,
        "token": null,
        "uid_file_owner": "admin",
        "note": "",
        "label": null,
        "displayname_file_owner": "admin",
        "path": "\/__sharedByAdmin",
        "item_type": "folder",
        "mimetype": "httpd\/unix-directory",
        "storage_id": "shared::\/__sharedByAdmin",
        "storage": 1,
        "item_source": 72003,
        "file_source": 72003,
        "file_parent": 103,
        "file_target": "\/__sharedByAdmin",
        "share_with": "tobi",
        "share_with_displayname": "Kaminsky Tobi",
        "mail_send": 0,
        "hide_download": 0
      },
```

This means I cannot simply display all received shares for a given folder, but first have to check if "share_with" is not the username.
As this behaviour changed, I expect that other apps on mobile/web/dav will also need to be changed…

File: lib/private/Share20/DefaultShareProvider.php (hunk 1 above)
Comment (daita) [reply lib/private/Share20/DefaultShareProvider.php]: note that I could swap back to select() as I filters duplicate results

**Comment by daita** [comment]
@tobiasKaminsky I think it is the main idea behind the edit: displaying all shares if you have reshares rights ? or am I missing something ?

**Comment by tobiasKaminsky** [comment]
> @tobiasKaminsky I think it is the main idea behind the edit: displaying all shares if you have reshares rights ? or am I missing something ?

But seeing myself as a share is kind of useless, or?

This is how it currently looks like: 
File is shared by admin with me (Kaminsky Tobi) and I additionally see myself as a sharee.


In my expectation it should be this way:
- with reshare permission see all sharees except your own
	- see what other users admin shared to
	- see your sharees 

So, if admin shared this files also to userA, userB and I shared it to friendA, friendB, I would expect that I get this returend via API:
- userA (read only as I am not allowed to edit this)
- userB (read only as I am not allowed to edit this)
- friendA (I can change this)
- friendB (I can change this)

but I do not see myself in this list

**Comment by daita** [comment]
Ok, wasn't clear who was the viewer, let me check what can be done

**Comment by daita** [comment]
@tobiasKaminsky can you try now ? 

_(Also, in my opinion, it should not be filtered by the API but by the client itself)_

**Comment by tobiasKaminsky** [comment]
> @tobiasKaminsky can you try now ?
> 
> _(Also, in my opinion, it should not be filtered by the API but by the client itself)_

works for me now as expected :tada:

**Comment by MorrisJobke** [comment]
@rullzer Mind to review?

File: lib/private/Share20/DefaultShareProvider.php (hunk 1 above)
Comment (MorrisJobke) [reply lib/private/Share20/DefaultShareProvider.php]: So this can go then? Or should we leave it ?

**Review by rullzer — APPROVED** [review]
Let do this then. :elephant: saw @tobiasKaminsky test it and code makes sese

**Pull request merged**